### PR TITLE
perf(streaming): add the missed read prefix hint on state table

### LINKED ISSUE #8541: perf(streaming): add `read_prefix_len_hint` in distinct agg table
As title.
found by @Little-Wallace

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

**Could impact nexmark q15(agg distinct) https://github.com/risingwavelabs/risingwave/issues/7350, q16(agg distinct) https://github.com/risingwavelabs/risingwave/issues/7271 and q102(dyn filter)**
## What's changed and what's your intention?
**should NOT merge before** https://github.com/risingwavelabs/risingwave/pull/8544, so that we can see the plan's changes.

- add `read_prefix_len_hint` in the `build` function of the `TableCatalogBuilder ` as a required field.
- add `read_prefix_len_hint for 
  - agg distinct table, fix #8541 
  - watermark filter state table
  - group topn state table
  - source state table
- 
## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Comment by st1page** [comment]
Could impact nexmark q15(agg distinct), q16(agg distinct) and q102(dyn filter)

File: src/frontend/src/optimizer/plan_node/generic/agg.rs [hunk 1]
```diff
@@ -196,33 +196,56 @@ impl<PlanRef: stream::StreamPlanRef> Agg<PlanRef> {
...
             for &idx in &self.group_key {
-                add_column(idx, Some(OrderType::ascending()), false);
+                add_column(
+                    idx,
+                    Some(OrderType::ascending()),
+                    false,
+                    &mut internal_table_catalog_builder,
+                );
             }
+            let read_prefix_len_hint = internal_table_catalog_builder.get_current_pk_len();
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/plan_node/generic/agg.rs:229]: Is this the reason we cannot directly capture the builder in the closure?

File: proto/catalog.proto [hunk 2]
```diff
@@ -160,6 +160,8 @@ message Table {
   repeated int32 value_indices = 20;
   string definition = 21;
   HandleConflictBehavior handle_pk_conflict_behavior = 22;
+  // Anticipated read prefix pattern (number of fields) for the table, which can be utilized
+  // for implementing the table's bloom filter or other storage optimization techniques.
   uint32 read_prefix_len_hint = 23;
```
Comment (BugenZhao) [inline proto/catalog.proto:165]: Do you think it's worthwhile to add this info to `distsql` explain?

**Review by BugenZhao — APPROVED** [review]

File: proto/catalog.proto (hunk 2 above)
Comment (BugenZhao) [reply proto/catalog.proto:165]: Seems we get the same idea. 😄 #8544

File: src/frontend/src/optimizer/plan_node/generic/agg.rs (hunk 1 above)
Comment (st1page) [reply src/frontend/src/optimizer/plan_node/generic/agg.rs:229]: Yes :hot_face: I was just tired and not sure if there could be better method.

**Review by richardchien — APPROVED** [review]
LGTM!

**Pull request merged**